prow/cmd/image-builder/buildcontroller.go

```diff
@@ -253,110 +266,50 @@ func (r *buildReconciler) createPVC(ctx context.Context, ibPod *corev1.Pod) (*co
 	if err != nil {
 		return errors.Wrap(err, "define clonerefs pod")
 	}
 	r.buildPods = append(r.buildPods, buildPod{pod: clonerefsPod, buildGroup: "clonerefs"})
 
-	// Next pods build the targets
-	for i, target := range r.options.targets.Strings() {
-
-		// Base configuration of build pod
-		var buildGroup string
-		pod := corev1.Pod{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      r.getBuildPodName(ibPod, target),
-				Namespace: ibPod.Namespace,
-			},
-			Spec: corev1.PodSpec{
-				RestartPolicy: corev1.RestartPolicyNever,
-				Volumes: []corev1.Volume{
-					{
-						Name: dockerConfigVolume,
-						VolumeSource: corev1.VolumeSource{
-							Secret: &corev1.SecretVolumeSource{
-								SecretName: r.options.dockerConfigSecret,
-							},
-						},
-					},
-				},
-			},
-		}
+	var variants []buildVariant
 
-		// Base configuration of kaniko container
-		kanikoContainer := corev1.Container{
-			Name:  "kaniko",
-			Image: r.options.kanikoImage,
-			Args: []string{
-				"--skip-unused-stages",
-				"--context=/code",
-				fmt.Sprintf("--dockerfile=%s", r.options.dockerfile),
-				fmt.Sprintf("--target=%s", target),
-			},
-			VolumeMounts: []corev1.VolumeMount{
-				{
-					Name:      codeVolume,
-					MountPath: "/code",
-					SubPath:   "code",
-				},
-				{
-					Name:      dockerConfigVolume,
-					MountPath: "/kaniko/.docker",
-				},
-			},
-			Resources: corev1.ResourceRequirements{
-				Requests: corev1.ResourceList{
-					corev1.ResourceCPU:    qtyZero,
-					corev1.ResourceMemory: qtyZero,
-				},
-			},
-		}
-
-		// Add destinations
-		destinations, err := r.defineDestinations(target)
+	if r.options.buildDefintion != "" {
+		// Get variants for build-definition case
+		variants, err = r.getVariants()
 		if err != nil {
-			return errors.Wrap(err, "construct destinations")
-		}
-		kanikoContainer.Args = append(kanikoContainer.Args, destinations...)
-
-		// Add kaniko args
-		kanikoContainer.Args = append(kanikoContainer.Args, r.options.kanikoArgs.Strings()...)
-
-		// Set caching options
-		if r.options.cacheRegistry != "" {
-			kanikoContainer.Args = append(
-				kanikoContainer.Args,
-				"--cache=true",
-				fmt.Sprintf("--cache-repo=%s", r.options.cacheRegistry),
-			)
+			return errors.Wrap(err, "get build variants")
 		}
+		r.log.Infof("Selected variants for this build: %+v", variants)
+	} else {
+		// Empty variant for the regular build case
+		variants = append(variants, buildVariant{name: nil, buildArgs: nil})
+	}
 
-		if i == 0 && r.options.cacheRegistry != "" {
-			buildGroup = "createCache"
-		} else {
-			buildGroup = "parallelBuild"
-		}
+	// Next pods build the targets for the variants
+	for _, variant := range variants {
+		for i, target := range r.options.targets.Strings() {
 
-		// Append build container to build pod
-		pod.Spec.Containers = append(pod.Spec.Containers, kanikoContainer)
+			pod, err := r.definePodForTarget(ibPod, target, variant.name, variant.buildArgs)
```

Comment:
Pass the entire `variant`?
```suggestion
			pod, err := r.definePodForTarget(ibPod, target, variant)
```

Reply:
done 👍